# Evaluation

---



**Objective**:

Evaluate the models to ensure they meet the business objectives. This might involve deploying them in a test environment.

**Tasks:**

1. Evaluate results.
2. Review the process.
3. Determine the next steps.

First, we'll evaluate the Random Forest and Logistic Regression models' performances on the test data to understand how well they generalize to new data.

In [2]:
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)
from google.colab import files

uploaded = files.upload()

Saving heart.csv to heart.csv


In [3]:
import pandas as pd
data = pd.read_csv('heart.csv')

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Splitting the data into features and target variable
X = data.drop(columns=['output'])
y = data['output']

# Feature Scaling (Normalization)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Splitting the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape

((242, 13), (61, 13))

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Initialize and train the Logistic Regression model
log_reg = LogisticRegression(random_state=42)
log_reg.fit(X_train, y_train)

# Predictions on the training set
y_train_pred = log_reg.predict(X_train)

# Evaluate the model's performance on the training data
accuracy = accuracy_score(y_train, y_train_pred)
class_report = classification_report(y_train, y_train_pred)

accuracy, class_report

(0.8636363636363636,
 '              precision    recall  f1-score   support\n\n           0       0.89      0.80      0.84       109\n           1       0.85      0.92      0.88       133\n\n    accuracy                           0.86       242\n   macro avg       0.87      0.86      0.86       242\nweighted avg       0.87      0.86      0.86       242\n')

In [7]:
from sklearn.ensemble import RandomForestClassifier
import xgboost as xgb

# Initialize and train the Random Forest model
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
y_train_rf = rf_model.predict(X_train)
rf_accuracy = accuracy_score(y_train, y_train_rf)

# Initialize and train the XGBoost model
xgb_model = xgb.XGBClassifier(random_state=42, eval_metric='logloss')
xgb_model.fit(X_train, y_train)
y_train_xgb = xgb_model.predict(X_train)
xgb_accuracy = accuracy_score(y_train, y_train_xgb)

rf_accuracy, xgb_accuracy

(1.0, 1.0)

In [8]:
# Predictions on the test set using Random Forest and Logistic Regression models
y_test_rf = rf_model.predict(X_test)
y_test_log_reg = log_reg.predict(X_test)

# Evaluate the models' performance on the test data
rf_test_accuracy = accuracy_score(y_test, y_test_rf)
log_reg_test_accuracy = accuracy_score(y_test, y_test_log_reg)

rf_test_accuracy, log_reg_test_accuracy



(0.8360655737704918, 0.8524590163934426)

The models' performances on the test data are as follows:

**Random Forest Classifier:**
* Approximately 83.61% accuracy
* Logistic Regression: Approximately 85.25% accuracy

Both models are performing well on the test data, with the Logistic Regression model slightly outperforming the Random Forest Classifier.

**Review Process:**

* We followed the CRISP-DM methodology, starting with understanding the business problem, exploring and preparing the data, modeling, and finally evaluating the models.
* The data was clean without missing values, which simplified the data preparation process.
* We normalized the features to ensure consistent scales.
* We evaluated multiple models, with Logistic Regression and Random Forest yielding promising results on both training and test data.

**Determine Next Steps:**

1. We can further tune hyperparameters for both models to potentially improve accuracy.
2. We can explore other algorithms or ensemble techniques.
3. Feature engineering could be explored more to derive new features or refine existing ones.
4. Once finalized, the model can be deployed in a real-world environment for predictions.